File: src/0304_fundamentals_of_predictive_modelling/04_assignment/assignment_01.py

# BACKGROUND: 
#   
#   The data for modeling contains information on Selling price of each house in million 
#   Rs. It also contains Carpet area in square feet, Distance from nearest metro station 
#   and Number of schools within 2 km distance. The data has 198 rows and 5 columns.



import numpy as np
import pandas as pd
import seaborn as sns
import statsmodels.api as sm

from patsy import dmatrices
from scipy import stats
from sklearn.model_selection import train_test_split
from statsmodels.stats.outliers_influence import variance_inflation_factor
from statsmodels.formula.api import ols
from statsmodels.graphics.regressionplots import influence_plot
from statsmodels.stats.diagnostic import lilliefors



# %% 1 - Import House Price Data. Check the structure of the data.
data = pd.read_csv("../../../data/0304_fundamentals_of_predictive_modelling/assignment/House Price Data.csv")

data.head()
#    Houseid  Price  Area  Distance  Schools
# 0        1  24.74  1036      3.22        2
# 1        2  20.15  1030      4.33        3
# 2        3  25.98  1046      1.94        3
# 3        4  20.10   950      2.45        2
# 4        5  23.03   952      2.47        2

data.info()
# Data columns (total 5 columns):
#  #   Column    Non-Null Count  Dtype  
# ---  ------    --------------  -----  
#  0   Houseid   198 non-null    int64  
#  1   Price     198 non-null    float64
#  2   Area      198 non-null    int64  
#  3   Distance  198 non-null    float64
#  4   Schools   198 non-null    int64  
# dtypes: float64(2), int64(3)
# memory usage: 7.9 KB



# %% 2 - Split the data into Training (80%) and Testing (20%) data sets
train, test = train_test_split(data, test_size = 0.2, random_state = 0)



# %% 3 - Build a regression model on training data to estimate selling price of a House.
model = ols('Price ~ Area + Distance + Schools', data = train).fit()
model.summary()
# """
#                             OLS Regression Results                            
# ==============================================================================
# Dep. Variable:                  Price   R-squared:                       0.797
# Model:                            OLS   Adj. R-squared:                  0.794
# Method:                 Least Squares   F-statistic:                     202.1
# Date:                Sat, 16 Dec 2023   Prob (F-statistic):           3.57e-53
# Time:                        16:00:24   Log-Likelihood:                -350.50
# No. Observations:                 158   AIC:                             709.0
# Df Residuals:                     154   BIC:                             721.3
# Df Model:                           3                                         
# Covariance Type:            nonrobust                                         
# ==============================================================================
#                  coef    std err          t      P>|t|      [0.025      0.975]
# ------------------------------------------------------------------------------
# Intercept     -8.6063      1.933     -4.452      0.000     -12.425      -4.788
# Area           0.0333      0.002     15.008      0.000       0.029       0.038
# Distance      -1.9165      0.190    -10.073      0.000      -2.292      -1.541
# Schools        1.5059      0.428      3.515      0.001       0.659       2.352
# ==============================================================================
# Omnibus:                        7.437   Durbin-Watson:                   1.986
# Prob(Omnibus):                  0.024   Jarque-Bera (JB):                7.606
# Skew:                          -0.537   Prob(JB):                       0.0223
# Kurtosis:                       2.977   Cond. No.                     1.14e+04
# ==============================================================================

# Notes:
# [1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
# [2] The condition number is large, 1.14e+04. This might indicate that there are
# strong multicollinearity or other numerical problems.
# """



# %% 4 - List down significant variables and interpret their regression coefficients.
model.params
# Intercept   -8.606252
# Area         0.033306
# Distance    -1.916504
# Schools      1.505935
# dtype: float64

# ==============================================================================
#                  coef    std err          t      P>|t|      [0.025      0.975]
# ------------------------------------------------------------------------------
# Intercept     -8.6063      1.933     -4.452      0.000     -12.425      -4.788
# Area           0.0333      0.002     15.008      0.000       0.029       0.038
# Distance      -1.9165      0.190    -10.073      0.000      -2.292      -1.541
# Schools        1.5059      0.428      3.515      0.001       0.659       2.352
# ==============================================================================

# for every unit increase in Area, the Price increases by 0.033306
# for every unit increase in Distance, the Price decreases by 1.916504
# for every unit increase in Schools, the Price increases by 1.505935



# %% 5 - What is the R2 and adjusted R2 of the model? Give interpretation.
# R-squared:                       0.797
# Adj. R-squared:                  0.794

# R-squared gives the proportion of the variation in the dependent variable that is 
# explained by the predictors / independent variables. 

# Adjusted R-squared is the value of R-squared adjusted for the number of predictors 
# in the model.

# in the case of R-squared, the value 0.797 implies that 79% - 80% of the variation in 
# Price is explained by the model (and 20% - 21% is unexplained variation)



# %% 6 - Is there a multicollinearity problem? If yes, do the necessary steps to remove it.
y, X = dmatrices('Price ~ Area + Distance + Schools', data = train, return_type = "dataframe")
pd.Series([variance_inflation_factor(X.values, i) for i in range(X.shape[1])], index = X.columns)
# Intercept    116.312936
# Area           1.569294
# Distance       1.055975
# Schools        1.631114
# dtype: float64

# as all calculated values of vif for each predictor are less than 5, there does not 
# appear to be a multicolinearity problem



# %% 7 - Are there any influential observations in the data?
influence = model.get_influence()
influence.summary_frame()
#      dfb_Intercept  dfb_Area  ...  student_resid    dffits
# 131       0.356091 -0.391085  ...      -1.540450 -0.425236
# 173       0.000168 -0.000555  ...       0.019284  0.002136
# 26        0.075654 -0.015935  ...      -0.804566 -0.163596
# 61       -0.059823  0.097184  ...      -1.076761 -0.149354
# 145      -0.015229  0.002613  ...      -0.608542 -0.082478
# ..             ...       ...  ...            ...       ...
# 67       -0.001200 -0.022123  ...       0.241596  0.055881
# 192       0.005931 -0.009491  ...       0.283316  0.034042
# 117      -0.075854  0.149698  ...       1.207080  0.205397
# 47       -0.003956 -0.000854  ...       0.197919  0.020719
# 172      -0.071516  0.058758  ...       0.933468  0.114178

# We can see from the above summary that there are many influential observations in 
# the data.

fig = influence_plot(model, criterion = 'Cooks')



# %% 8 - Can we assume that errors follow ‘Normal’ distribution?
train = train.assign(pred = pd.Series(model.fittedvalues))
train = train.assign(resi = pd.Series(model.resid))

train.head()
#      Houseid  Price  Area  Distance  Schools       pred      resi
# 131      132  33.41  1345      2.07        3  36.740684 -3.330684
# 173      174  27.76  1066      1.93        3  27.716678  0.043322
# 26        27  24.00  1139      4.21        3  25.778371 -1.778371
# 61        62  23.01  1006      2.09        3  25.411689 -2.401689
# 145      146  19.52   990      3.39        2  20.881406 -1.361406

fig = sm.graphics.qqplot(train.resi, line = '45', fit = True)

# The q-q plot appears to be reasonably normal, but deviating from normality at the tails.
# We should perform the standard tests for normality to confirm.

stats.shapiro(train.resi)
# ShapiroResult(statistic=0.9755853414535522, pvalue=0.006665823049843311)

# as the p-value is 0.006665823049843311 we reject the null hypothesis that the residuals 
# are drawn from normally distributed data. 

lilliefors(train.resi)
# (0.06475925533004229, 0.15433881296699287)

# as the p-value is 0.15433881296699287 we fail to reject the null hypothesis that the 
# residuals are drawn from normally distributed data.

# the above tests give conflicting results, but looking at the q-q plot we can see that
# both the tails deviate from normality, and hence the Shapiro-Wilk test will likely give 
# a p-value < 0.05 as it is concerned primarily with tails. We accept the outcome of the 
# Kolmogorov-Smirnov test.



# %% 9 - Is there a Heteroscedasticity problem? Check using residual vs. predictor plots.
train.plot.scatter(x = 'pred', y = 'resi')

# residuals appear to be reasonably randomly distributed, which would indicate 
# homoscedasticity, and hence no heteroscedasticity problem.



# %% 10 - Calculate the RMSE for the Training and Testing data.
RMSE_train = np.sqrt((train.resi ** 2).mean())
RMSE_train
# 2.224293038804313

test = test.assign(pred = pd.Series(model.predict(test)))
test = test.assign(resi = pd.Series(test.Price - test.pred))

test.head()
#      Houseid  Price  Area  Distance  Schools       pred      resi
# 18        19  27.94  1111      2.69        2  26.252961  1.687039
# 168      169  21.44   950      2.45        2  21.350688  0.089312
# 63        64  25.03  1033      3.21        3  24.164461  0.865539
# 175      176  17.11   916      3.15        2  18.876738 -1.766738
# 71        72  20.68   967      3.64        2  19.636247  1.043753

RMSE_test = np.sqrt((test.resi ** 2).mean())
RMSE_test
# 2.116735309570193

((RMSE_train - RMSE_test) / RMSE_train) * 100
# 4.835591684984935

# the difference between the train and test RMSE is around 4% - 5% - which is acceptable 
# (a difference of around 10% is usually the ideal). Also of note is that the test RMSE 
# is less than the train RMSE, which would indicate that the 
# error has decreased going from train to test.

File: scripts/0304_fundamentals_of_predictive_modelling/04_assignment/assignment_01.R

# BACKGROUND: 
#   
#   The data for modeling contains information on Selling price of each house in million 
#   Rs. It also contains Carpet area in square feet, Distance from nearest metro station 
#   and Number of schools within 2 km distance. The data has 198 rows and 5 columns.

library(car)
library(caret)
library(nortest)



set.seed(1337)



# 1 - Import House Price Data. Check the structure of the data.
data <- read.csv("./data/0304_fundamentals_of_predictive_modelling/04_assignment/House\ Price\ Data.csv", header = TRUE)

head(data)
#   Houseid Price Area Distance Schools
# 1       1 24.74 1036     3.22       2
# 2       2 20.15 1030     4.33       3
# 3       3 25.98 1046     1.94       3
# 4       4 20.10  950     2.45       2
# 5       5 23.03  952     2.47       2
# 6       6 21.02  967     3.64       2

str(data)
# 'data.frame':	198 obs. of  5 variables:
# $ Houseid : int  1 2 3 4 5 6 7 8 9 10 ...
# $ Price   : num  24.7 20.1 26 20.1 23 ...
# $ Area    : int  1036 1030 1046 950 952 967 825 1162 1066 1084 ...
# $ Distance: num  3.22 4.33 1.94 2.45 2.47 3.64 1.49 2.26 1.93 1.47 ...
# $ Schools : int  2 3 3 2 2 2 2 3 3 2 ...



# 2 - Split the data into Training (80%) and Testing (20%) data sets
index <- createDataPartition(data$Price, p = 0.8, list = FALSE)
train <- data[index, ]
test  <- data[-index, ]



# 3 - Build a regression model on training data to estimate selling price of a House.
model <- lm(Price ~ Area + Distance + Schools, data = train)
summary(model)
# Call:
# lm(formula = Price ~ Area + Distance + Schools, data = train)
# 
# Residuals:
#     Min      1Q  Median      3Q     Max 
# -6.1814 -1.3303  0.1795  1.6341  4.6410 
# 
# Coefficients:
#              Estimate Std. Error t value Pr(>|t|)    
# (Intercept) -9.339229   1.961860  -4.760 4.38e-06 ***
# Area         0.034376   0.002258  15.223  < 2e-16 ***
# Distance    -1.751836   0.191119  -9.166 2.71e-16 ***
# Schools      1.170730   0.408580   2.865  0.00474 ** 
# ---
# Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
# 
# Residual standard error: 2.244 on 156 degrees of freedom
# Multiple R-squared:  0.7786,	Adjusted R-squared:  0.7743 
# F-statistic: 182.9 on 3 and 156 DF,  p-value: < 2.2e-16



# 4 - List down significant variables and interpret their regression coefficients.
model$coefficients
# (Intercept)        Area    Distance     Schools 
# -9.33922947  0.03437617 -1.75183599  1.17072991   

# Coefficients:
#              Estimate Std. Error t value Pr(>|t|)    
# (Intercept) -8.587180   1.919831  -4.473 1.48e-05 ***
# Area         0.034074   0.002238  15.222  < 2e-16 ***
# Distance    -1.945728   0.180181 -10.799  < 2e-16 ***
# Schools      1.237763   0.414879   2.983  0.00331 ** 

# for every unit increase in Area, the Price increases by 0.03437617
# for every unit increase in Distance, the Price decreases by 1.75183599
# for every unit increase in Schools, the Price increases by 1.17072991



# 5 - What is the R2 and adjusted R2 of the model? Give interpretation.
# R-squared:           0.7786
# Adjusted R-squared:  0.7743

# R-squared gives the proportion of the variation in the dependent variable that is 
# explained by the predictors / independent variables. 

# Adjusted R-squared is the value of R-squared adjusted for the number of predictors 
# in the model.

# in the case of R-squared, the value 0.7786 implies that 77% - 78% of the variation in 
# Price is explained by the model (and 22% - 23% is unexplained variation)



# 6 - Is there a multicollinearity problem? If yes, do the necessary steps to remove it.
vif(model)
#     Area Distance  Schools 
# 1.559863 1.033669 1.590641 

# as all calculated values of vif for each predictor are less than 5, there does not 
# appear to be a multicolinearity problem



# 7 - Are there any influential observations in the data?
summary(influence.measures(model))
# Potentially influential observations of
#          lm(formula = Price ~ Area + Distance + Schools, data = train) :
#   
#     dfb.1_ dfb.Area dfb.Dstn dfb.Schl dffit   cov.r   cook.d hat    
# 17  -0.04   0.26    -0.09    -0.45     0.47    1.04    0.05   0.08_*
# 32  -0.33   0.12     0.40     0.14    -0.53_*  0.88_*  0.07   0.04  
# 33   0.39  -0.42     0.07     0.16    -0.46    1.04    0.05   0.08_*
# 35  -0.05   0.11     0.10    -0.20     0.25    1.11_*  0.02   0.09_*
# 65  -0.30   0.11     0.37     0.13    -0.49_*  0.90_*  0.06   0.04  
# 66   0.41  -0.44     0.07     0.17    -0.48    1.03    0.06   0.08_*
# 98  -0.34   0.13     0.41     0.14    -0.55_*  0.86_*  0.07   0.04  
# 99   0.41  -0.43     0.07     0.16    -0.47    1.04    0.05   0.08_*
# 116 -0.04   0.27    -0.10    -0.48     0.50_*  1.03    0.06   0.08_*
# 131 -0.29   0.11     0.36     0.12    -0.47    0.91_*  0.05   0.04  
# 132  0.39  -0.41     0.07     0.16    -0.45    1.04    0.05   0.08_*
# 134 -0.05   0.12     0.12    -0.22     0.28    1.10_*  0.02   0.09_*
# 198  0.38  -0.40     0.06     0.15    -0.44    1.05    0.05   0.08_*

train[cooks.distance(model) > (4 / nrow(train)), ]
#     Houseid Price Area Distance Schools
# 17       17 26.75 1028     2.23       1
# 18       18 34.97 1072     0.10       3
# 32       32 17.73  929     0.70       2
# 33       33 33.36 1345     2.07       3
# 65       65 18.10  929     0.70       2
# 66       66 33.20 1345     2.07       3
# 98       98 17.53  929     0.70       2
# 99       99 33.25 1345     2.07       3
# 116     116 26.99 1028     2.23       1
# 131     131 18.35  929     0.70       2
# 132     132 33.41 1345     2.07       3
# 150     150 35.49 1072     0.10       3
# 164     164 18.98  929     0.70       2
# 183     183 34.93 1072     0.10       3
# 198     198 33.49 1345     2.07       3

# We can see from the above summary that there are as many as 15 influential 
# observations in the training data.

influencePlot(model, main = "Influence Plot", sub = "Circle size is proportioal to Cook's Distance")
#        StudRes        Hat      CookD
# 32  -2.7703899 0.03506313 0.06686166
# 35   0.8101086 0.08992431 0.01624741
# 98  -2.8678549 0.03506313 0.07140771
# 134  0.8944953 0.08992431 0.01979031



# 8 - Can we assume that errors follow ‘Normal’ distribution?
train$pred <- fitted(model)
train$resi <- residuals(model)

head(train)
#   Houseid Price Area Distance Schools     pred       resi
# 1       1 24.74 1036     3.22       2 22.97503  1.7649656
# 3       3 25.98 1046     1.94       3 26.73188 -0.7518761
# 4       4 20.10  950     2.45       2 21.36760 -1.2675972
# 5       5 23.03  952     2.47       2 21.40131  1.6286872
# 6       6 21.02  967     3.64       2 19.86731  1.1526927
# 7       7 17.44  825     1.49       2 18.75234 -1.3123380

# q-q plot of resi
qqnorm(train$resi, col = "cadetblue")
qqline(train$resi, col = "cadetblue", lwd = 2)
# The q-q plot appears to be reasonably normal, but deviating from normality at the tails.
# We should perform the standard tests for normality to confirm.

shapiro.test(train$resi)
#         Shapiro-Wilk normality test
# 
# data:  train$resi
# W = 0.98159, p-value = 0.03149

# as the p-value is 0.03149 we reject the null hypothesis that the residuals are drawn 
# from normally distributed data. However, we would fail to reject the null hypothesis 
# at the 0.01 level of significance (99%)

lillie.test(train$resi)
#         Lilliefors (Kolmogorov-Smirnov) normality test
# 
# data:  train$resi
# D = 0.062922, p-value = 0.1261

# as the p-value is 0.1261 we fail to reject the null hypothesis that the residuals are 
# drawn from normally distributed data.

# the above tests give conflicting results, but looking at the q-q plot we can see that
# both the tails deviate from normality, and hence the Shapiro-Wilk test will likely give 
# a p-value < 0.05 as it is concerned primarily with tails. We accept the outcome of the 
# Kolmogorov-Smirnov test.



# 9 - Is there a Heteroscedasticity problem? Check using residual vs. predictor plots.
plot(train$pred, train$resi, col = "cadetblue")

# residuals appear to be reasonably randomly distributed, which would indicate 
# homoscedasticity, and hence no heteroscedasticity problem.



# 10 - Calculate the RMSE for the Training and Testing data.
RMSE_train <- sqrt(mean(train$resi ** 2))
RMSE_train
# 2.216225

test$pred <- predict(model, test)
test$resi <- test$Price - test$pred

head(test)
#    Houseid Price Area Distance Schools     pred      resi
# 2        2 20.15 1030     4.33       3 21.99497 -1.844969
# 12      12 34.10 1138     0.88       3 31.75143  2.348570
# 14      14 19.75  990     3.39       2 21.09592 -1.345918
# 24      24 30.85 1160     2.80       3 29.14418  1.705819
# 37      37 20.18  950     2.45       2 21.36760 -1.187597
# 51      51 34.92 1072     0.10       3 30.84903  4.070965

RMSE_test <- sqrt(mean(test$resi ** 2))
RMSE_test
# 2.17643

# both values of RMSE are consistent, which is a strong indicator of the stability of 
# the model.

((RMSE_train - RMSE_test) / RMSE_train) * 100
# 1.795594

# the difference between the train and test RMSE is around 1% - 2% - which is acceptable 
# (a difference of around 10% is usually the ideal). Also of note is that the test RMSE 
# is less than the train RMSE, which would indicate that the error has decreased going
# from train to test.
